multicast_expert/utils.py

```diff
@@ -1,17 +1,23 @@
-import socket
-from typing import List, Dict, Optional, Union, Tuple
+from __future__ import annotations
+
 import ipaddress
 import platform
+import socket
+from typing import Tuple, Union
 
 import netifaces
 
+# docsig: disable
+
 is_windows = platform.system() == "Windows"
 is_mac = platform.system() == "Darwin"
 
 
 # Type which can represent an IPv4 or an IPv6 address.
 # For IPv4, address is a tuple of IP address (str) and port number.
 # For IPv6, address is a tuple of IP address (str), port number, flow info (int), and scope ID (int).
+# NOTE: Since we still support python 3.9, we cannot use new style type annotations here. We have to use the
+# old ones which still work in 3.9
 IPv4Or6Address = Union[Tuple[str, int], Tuple[str, int, int, int]]
```

Comment:
I think you can use:

```suggestion
IPv4Or6Address = Union[tuple[str, int], tuple[str, int, int, int]]
```

Python 3.9 added support for using `tuple` instead of `Tuple`, and you can mix new-style `tuple` annotations with old-style `Union` annotations. Python 3.10 added support for new-style `X | Y` union annotations, but Ruff won't ask you to add them because `target-version` is set to `3.9`.

Reply:
ahh I see, I didn't realize it was like this!